### PR TITLE
Image uploads resize and use correct endpoints

### LINKED ISSUE #275: Uploaded images should automatically be downscaled

### PR DESCRIPTION
**Pull request by kmcnellis**
fixes #275

### CODE REVIEW

File: server/api/course/course.controller.js [hunk 1]
```diff
@@ -63,25 +66,22 @@ function removeEntity(res) {
   };
 }
 
-function saveEventInfoImage(files, fields, cb){
+function saveCourseImage(files, fields, cb){
```
Comment (seveibar) [inline server/api/course/course.controller.js:70]: wow how did this one slip through haha good fix

File: server/api/eventinfo/eventinfo.controller.js [hunk 2]
```diff
@@ -67,22 +70,20 @@ function removeEntity(res) {
...
-    asyncTasks.push( (callback) => {fs.exists(path, (exists) => {
-      mkdirp(path, function (err) {
-        if (err) console.error(err)
-        var is = fs.createReadStream(file.path);
-        var os = fs.createWriteStream(destPath);
-        is.pipe(os);
-        callback();
+    var path = config.imageUploadPath  + 'eventInfoImages' + '/';
+    asyncTasks.push( (callback) => {
+      var imagePath = imageUpload.saveImage(file, path, function(err) {
```
Comment (seveibar) [inline server/api/eventinfo/eventinfo.controller.js:81]: I like this modularization- we should have been doing this from the beginning

File: server/api/eventinfo/eventinfo.controller.js [hunk 3]
```diff
@@ -185,6 +186,64 @@ exports.create = function(req, res) {
...
+  if (req.query.size){
+      var extension = path.extname(name);
+      var id = name.substring(0,name.length-extension.length);
+
+    if (req.query.size === 'preview'){
+      name = id + "-preview" + extension;
+    }
+    else if (req.query.size === 'large'){
+      name = id + "-large" + extension;
+    }
```
Comment (seveibar) [inline server/api/eventinfo/eventinfo.controller.js]: cool i like this method

File: server/api/eventinfo/eventinfo.controller.js [hunk 4]
```diff
@@ -185,6 +186,64 @@ exports.create = function(req, res) {
...
+    source = config.imageUploadPath + 'eventInfoImages/' + name;
+    imgPath = path.join(__dirname, "../../../", source);
+    try {
+        // Query the entry
+        stats = fs.lstatSync(imgPath);
+        return res.sendFile(imgPath);
+    }
+    catch (e) {
+      // File not found, send nothing
+      return res.sendFile("");
```
Comment (seveibar) [inline server/api/eventinfo/eventinfo.controller.js]: ehhh :penguin: i think maybe this should 404

File: server/api/submission/submission.controller.js [hunk 5]
```diff
@@ -74,42 +76,23 @@ function removeEntity(res) {
...
-            var is = fs.createReadStream(file.path);
-            var os = fs.createWriteStream(destPath);
-            is.pipe(os);
-            callback();
-          });
-        });
+      var imagePath = imageUpload.saveImage(file, path, function(err) {
+        callback(err)
+      });
+      console.log("imagePath",imagePath)
```
Comment (seveibar) [inline server/api/submission/submission.controller.js]: loggggg

File: server/api/submission/submission.controller.js [hunk 6]
```diff
@@ -74,42 +76,23 @@ function removeEntity(res) {
...
+      imagePaths.push("/api/submissions/image?name=" + imagePath + "&userId=" + fields.userId + "&eventId=" + fields.eventId);
       });
     });
-  });
   async.parallel(asyncTasks, (error, results) => {
     // TODO: Handle Error
+    console.log(error);
```
Comment (seveibar) [inline server/api/submission/submission.controller.js]: loog

File: server/api/submission/submission.controller.js [hunk 7]
```diff
@@ -232,8 +215,59 @@ exports.index = function(req, res) {
...
+    source = config.imageUploadPath + 'eventImages/' + req.query.userId + '/' + req.query.eventId + '/' + name;
+    imgPath = path.join(__dirname, "../../../", source);
+    try {
+        // Query the entry
+        stats = fs.lstatSync(imgPath);
+        return res.sendFile(imgPath);
+    }
+    catch (e) {
+      // File not found, send nothing
+      return res.sendFile("");
```
Comment (seveibar) [inline server/api/submission/submission.controller.js]: again :penguin: maybe should 404

**Comment by seveibar** [comment]
Other than the two logs and possibly lack of a 404 on missing images i think it's good to go- I think someone should do a quick runthrough tho and make sure it works on other systems (I can do it tonight if nobody gets around to it)

**Comment by kmcnellis** [comment]
The big thing I want to change next is to switch from connect multiparty to multer. It uses the same backend tech but allows us to control what end points  are allowed to upload files (to avoid spamming). However, it requires a slight client end change (instead of it sending files[0], files[1], etc, they have to be named just files), so I'm worried about an iOS/android api change. Not sure if it's def going to be breaking, but requires some testing first. Should be a simple change, though.

The lack of images instead of 404 is in the same boat- it was that way previously and I'm not sure if iOS or android handles a potential 404

**Comment by kmcnellis** [comment]
@seveibar Could you take another look, I moved downloads to a module and forgot to upload the image upload module last time

**Comment by seveibar** [comment]
yea the fireworks are coming soon ill take a look right after

**Comment by seveibar** [comment]
Sorry for the delay I'll take a look at this soon (should be tested)

**Comment by seveibar** [comment]
@kmcnellis you shouldn't upsize photos right? Is that desired behavior?

**Comment by seveibar** [comment]
great job- i've created an issue for the upsizing b/c i think we might want to discuss whether that is desired behavior or not (and is much less important)

**Comment by kmcnellis** [comment]
It's currently set to not upsize. There is a setting to allow it if it's easier on your side.

**Comment by seveibar** [comment]
hmm it seemed to upsize my image (i went to the network tab and visited the url)

I'm not sure if the actual size was increased as a result of the resizing.

**Comment by seveibar** [comment]
Okay I'm not going to create and issue because I'm not remotely concerned about this haha looks good!

**Comment by kmcnellis** [comment]
Actually, I don't remember right now if it was set to or not... It is def a option to enable/disable

**Comment by seveibar** [comment]
I don't really consider it a problem- resolutions that low are not present on modern smartphones

**Pull request merged**